presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -683,6 +685,36 @@ protected Scope visitQuery(Query node, Optional<Scope> scope)
                     .withParent(withScope)
                     .withRelationType(RelationId.of(node), queryBodyScope.getRelationType())
                     .build();
+
+            // create a multimap of tables in a query to columns used for each table
+            Multimap<QualifiedObjectName, String> tableColumnMap = analysis.getColumnReferenceFields().values().stream()
+                    .filter(fieldId -> {
+                        // first filter out column references that don't directly reference tables
+                        Node sourceNode = fieldId.getRelationId().getSourceNode();
```

Comment:
This looks like a hack. I would prefer to collect this information more explicitly in `Analysis`, just like we do `com.facebook.presto.sql.analyzer.Analysis#tables`. Or find all `Table` nodes in the AST...

This brings an interesting question though. Assuming I don't have right to select from `tab_b.d`, should the following query fail?
```
select * from tab_a where exists (select d from tab_b); -- column `d` can be pruned
```

Reply:
>  I would prefer to collect this information more explicitly in Analysis, just like we do `com.facebook.presto.sql.analyzer.Analysis#tables`

Sounds good to me.

Reply:
You could gather references to table fields via `ExpressionAnalyzer.Visitor#handleResolvedField`. You should be always able to access `Field` there.

Reply:
@findepi  Thanks for the suggestion.  That's a much better idea.  Also, I don't think we should worry about columns that can be pruned.  If a user doesn't want us checking access for a column, they shouldn't put it in the query.  In fact, it might be confusing for it not to fail if it fails for other queries.

Reply:
Pruning is an optimization. Semantically, the columns are read, so the query should fail. And that’s what the spec says, btw.